File: JavaScript/Programmers/Level 1/예산.js
// 최대한 많은 부서에 예산을 지급해줘야 함, 즉, 예산이 제일 적은 부서부터 차례대로 예산을 지급해줌, 그리디 알고리즘 적용

// function solution(d, budget) {
//     let sum = 0;
//     return d.sort((a, b) => a - b) // 각 부서의 예산 오름차순 정렬
//         .filter(b => { // sum에 각 부서의 예산을 더해나감
//         sum += b;
//         if (sum > budget) // sum이 budget보다 크다면 더이상 예산을 지급해줄 수 없으므로 false return
//             return false;
        
//         return true; // 그게 아니라면 return true
//     }).length; // 필터링 된 배열에는 예산을 지급해줄 수 있는 부서만 남게 되므로 필터링 된 배열의 크기 return
// }

// filter 메소드의 매개변수로 콜백함수가 들어오므로 중간에 멈추지 않고 배열 전체를 순회하면서 true인 요소들만 배열에 남김
// d를 오름차순 정렬했으므로 배열 전체를 순회할 필요 없이 for문을 사용하여 예산을 지급해주다가 budget보다 큰 순간 순회를 멈춤

function solution(d, budget) {
    d = d.sort((a, b) => a - b);
    let idx, sum = 0;
    for (idx = 0; idx < d.length; idx++) {
        sum += d[idx];
        if (sum > budget) // 배열의 모든 요소를 순회하지 않고 sum이 budget보다 큰 순간 break
            break;
    }
    
    return idx; // break 되었을 때 idx가 예산을 지급해줄 수 있는 부서의 개수가 됨
}

File: Python/Programmers/Level 1/예산.py
# def solution(d, budget):
#     d.sort() # 예산이 제일 적은 부서부터 budget 지원해야 최대로 지원해줄 수 있음
    
#     answer = 0
#     for money in d:
#         if budget >= money: # budget이 각 부서 예산보다 크거나 같다면
#             budget -= money # budget에서 각 부서 예산만큼 빼기
#             answer += 1 # 그때마다 count
    
#     return answer

#  def solution(d, budget):
#     d.sort()
#     while budget < sum(d): # budget이 각 부서의 신청 예산의 합보다 작다면
#         d.pop() # 제일 큰 값 pop
    
#     return len(d) # 남은 d의 길이가 budget을 지원해줄 수 있는 최대값

# 위 드는 while문에서 매 조건을 확인할때마다 sum 함수가 반복적으로 호출되므로 효율성이 떨어짐
def solution(d, budget):
    answer = 0
    for price in sorted(d): # 최대한 많은 부서를 지원해주기 위해 d를 오름차순 정렬 후 모든 요소 순회
        budget -= price # budget에서 price 차감
        if budget < 0: # budget이 0보다 작으면 더이상 지원해줄 수 없으므로 break
            break
        
        answer += 1 # budget이 0이상이면 아직 지원해줄 수 있으므로 count
    
    return answer
